### PR TITLE
Use Kubernetes label selector

### PR DESCRIPTION
**Pull request by baijum**
Fixes https://issues.redhat.com/browse/APPSVC-322

### Motivation

Make API familiar to users through adopting Kubernetes native APIs

### Changes

Use `LabelSelector` from `k8s.io/apimachinery/pkg/apis/meta/v1` package

### Testing

The existing tests would be fine. The mock object has been modified.

### CODE REVIEW

**Comment by sbose78** [comment]
FYI: It is fine to have breaking changes with respect to labels since UI isn't consuming it?

File: pkg/controller/servicebindingrequest/reconciler.go [hunk 1]
```diff
@@ -169,9 +169,9 @@ func checkSBR(sbr *v1alpha1.ServiceBindingRequest, log *log.Log) error {
 		log.Debug("Spec.ApplicationSelector.ResourceRef not found")
 
 		// Check if MatchLabels is present
```
Comment (DhritiShikhar) [inline pkg/controller/servicebindingrequest/reconciler.go]: The comment should be updated

File: pkg/apis/apps/v1alpha1/servicebindingrequest_types.go [hunk 2]
```diff
@@ -63,11 +63,11 @@ type BackingServiceSelector struct {
 // +k8s:openapi-gen=true
 type ApplicationSelector struct {
 	// +optional
-	MatchLabels map[string]string `json:"matchLabels,omitempty"`
-	Group       string            `json:"group,omitempty"`
-	Version     string            `json:"version"`
-	Resource    string            `json:"resource"`
-	ResourceRef string            `json:"resourceRef"`
+	LabelSelector *metav1.LabelSelector `json:",inline"`
```
Comment (DhritiShikhar) [inline pkg/apis/apps/v1alpha1/servicebindingrequest_types.go]: Why an extra comma in the beginning `,inline`?

Comment (baijum) [reply pkg/apis/apps/v1alpha1/servicebindingrequest_types.go]: The tag name for the field is irrelevant here. For example, if the tag is like this: `json:"labelSelector,inline"` the word `labelSelector` will not appear anywhere as the values are embedded.

There is an explanation about `omitempty` option which is somewhat similar to `inline` option.
https://golang.org/pkg/encoding/json/#Marshal

I can't find any explanation about `inline` anywhere. But to understand the behavior you can run this example: https://gist.github.com/baijum/e761bc8fb28ba3855e5d1cd118e9192b

Let me know if any other clarification required.

**Comment by baijum** [comment]
> FYI: It is fine to have breaking changes with respect to labels since UI isn't consuming it?

Since the existing YAML structure is supported, I think we can consider it as backward compatible.
I am embedding the label selector. Let me know if this is fine.

**Comment by akashshinde** [comment]
@sbose78 @baijum UI isn't really consuming `matchLabels` anywhere yet, so I assume it's safe to update this field.

**Comment by baijum** [comment]
I have changed the API to use a nested structure.

File: pkg/apis/apps/v1alpha1/servicebindingrequest_types.go [hunk 3]
```diff
@@ -63,11 +63,11 @@ type BackingServiceSelector struct {
 // +k8s:openapi-gen=true
 type ApplicationSelector struct {
 	// +optional
-	MatchLabels map[string]string `json:"matchLabels,omitempty"`
-	Group       string            `json:"group,omitempty"`
-	Version     string            `json:"version"`
-	Resource    string            `json:"resource"`
-	ResourceRef string            `json:"resourceRef"`
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+	Group         string                `json:"group,omitempty"`
```
Comment (isutton) [inline pkg/apis/apps/v1alpha1/servicebindingrequest_types.go]: Perhaps GroupVersionResource can be embedded here as well?

Comment (baijum) [reply pkg/apis/apps/v1alpha1/servicebindingrequest_types.go]: We can do that in a separate PR.

**Pull request merged**